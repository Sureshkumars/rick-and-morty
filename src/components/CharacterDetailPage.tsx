import React from "react";
import {
  Box,
  VStack,
  Image,
  Heading,
  Button,
  Stat,
  StatLabel,
  StatNumber,
  useToast,
  Divider,
  List,
  ListIcon,
  ListItem,
} from "@chakra-ui/core";
import { PageError } from "./PageError";
import { SkeletonLoader } from "./SkeletonLoader";
import useCharacterDetailService from "../hooks/useCharacterDetailsService";
import { FaCheckCircle } from "react-icons/fa";
import { useParams } from "react-router-dom";
interface IParams {
  characterId: string;
}
export const CharacterDetailPage: React.FC = () => {
  const { characterId }: IParams = useParams();
  const toast = useToast();
  //get character details
  const characterDetails = useCharacterDetailService(characterId);
  return (
    <Box>
      {characterDetails.status === "loading" && <SkeletonLoader />}

      {characterDetails.status === "loaded" && (
        <VStack spacing={8} marginTop="10">
          <Image
            src={characterDetails.payload.imageUrl}
            alt={characterDetails.payload.name}
            boxSize="200px"
            fallbackSrc="/character-fallback.jpeg"
          />
          <Box>
            <Heading mb={4}>{characterDetails.payload.name}</Heading>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>species</StatLabel>
                <StatNumber>{characterDetails.payload.species}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Gender</StatLabel>
                <StatNumber>{characterDetails.payload.gender}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Origin Location</StatLabel>
                <StatNumber>{characterDetails.payload.origin}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Last Known Location</StatLabel>
                <StatNumber>{characterDetails.payload.lastLocation}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Total Episodes</StatLabel>
                <StatNumber>
                  {characterDetails.payload.totalEpisodes}
                </StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <List spacing={5}>
                <Stat>
                  <StatLabel>Episode Details</StatLabel>
                </Stat>
                {characterDetails.payload.episode.map((data) => {
                  return (
                    <ListItem key={data.name}>
                      <ListIcon
                        as={FaCheckCircle}
                        color="green.500"
                        fontSize="sm"
                      />
                      {data.name} - {data.season} - was aired at{" "}
                      {data.airedDate} <Divider orientation="horizontal" />
                    </ListItem>
                  );
                })}
              </List>
            </Box>
            <Button
              size="lg"
              colorScheme="blue"
              mt="24px"
              marginBottom="10"
              onClick={() =>
                toast({
                  title: "Purchase Successfull.",
                  description: "Merchandise was clicked successfully",
                  status: "success",
                  duration: 2000,
                  isClosable: true,
                })
              }
            >
              Buy Merchandise
            </Button>
          </Box>
        </VStack>
      )}
      {characterDetails.status === "error" && <PageError />}
    </Box>
  );
};
