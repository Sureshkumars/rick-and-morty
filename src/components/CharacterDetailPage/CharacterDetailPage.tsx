import React from "react";
import {
  Box,
  VStack,
  Image,
  Heading,
  Button,
  Stat,
  StatLabel,
  StatNumber,
  useToast,
  Divider,
  List,
  ListIcon,
  ListItem,
} from "@chakra-ui/core";
import { ErrorMessage } from "../ErrorMessage/ErrorMessage";
import { SkeletonLoader } from "../SkeletonLoader/SkeletonLoader";
import { FaCheckCircle } from "react-icons/fa";
import { useParams } from "react-router-dom";
import { ERROR_MESSAGE } from "../../constants/constants";
import useCharacterDetailService from "../../hooks/useCharacterDetailsService";

interface IParams {
  characterId: string;
}
export const CharacterDetailPage: React.FC = () => {
  const { characterId }: IParams = useParams();
  const toast = useToast();
  //Custom Hook to get character details
  const characterDetails = useCharacterDetailService(characterId);

  return (
    <Box>
      {characterDetails.status === "loading" && <SkeletonLoader />}

      {characterDetails.status === "loaded" && (
        <VStack spacing={8} marginTop="10">
          <Image
            src={characterDetails.payload.imageUrl}
            alt={characterDetails.payload.name}
            boxSize="200px"
            fallbackSrc="/images/character-fallback.jpeg"
          />
          <Box>
            <Heading mb={4}>{characterDetails.payload.name}</Heading>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>species</StatLabel>
                <StatNumber>{characterDetails.payload.species}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Gender</StatLabel>
                <StatNumber>{characterDetails.payload.gender}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Status</StatLabel>
                <StatNumber>{characterDetails.payload.status}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Origin Location</StatLabel>
                <StatNumber>{characterDetails.payload.origin}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Last Known Location</StatLabel>
                <StatNumber>{characterDetails.payload.lastLocation}</StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <Stat>
                <StatLabel>Total Episodes</StatLabel>
                <StatNumber>
                  {characterDetails.payload.totalEpisodes}
                </StatNumber>
              </Stat>
            </Box>
            <Box marginTop="1.25rem" padding="1rem" borderWidth="2px">
              <List spacing={5}>
                <Stat>
                  <StatLabel>Episode Details</StatLabel>
                </Stat>
                {characterDetails.payload.episode.map((data, i) => {
                  return (
                    <ListItem key={i}>
                      <ListIcon
                        as={FaCheckCircle}
                        color="green.500"
                        fontSize="sm"
                      />
                      {data.name} - {data.season} - was aired at{" "}
                      {data.airedDate} <Divider orientation="horizontal" />
                    </ListItem>
                  );
                })}
              </List>
            </Box>
            <Button
              aria-label="buy-merchandise"
              size="lg"
              colorScheme="blue"
              mt="24px"
              marginBottom="10"
              onClick={() =>
                toast({
                  title: "Purchase Successfull.",
                  description: "Merchandise was clicked successfully",
                  status: "success",
                  duration: 4000,
                  isClosable: true,
                })
              }
            >
              Buy Merchandise
            </Button>
          </Box>
        </VStack>
      )}
      {characterDetails.status === "error" && (
        <ErrorMessage message={ERROR_MESSAGE} showHomeCTA={true} />
      )}
    </Box>
  );
};
