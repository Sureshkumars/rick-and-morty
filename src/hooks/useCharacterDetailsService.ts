import { useEffect, useState } from "react";
import { CharacterDetails } from "../types/characterDetail";
import { Service } from "../types/service";
import { CHARACTER_API_ENDPOINT } from "../constants/constants";
const characterApiEndPoint = CHARACTER_API_ENDPOINT;

const useCharacterDetailService = (characterId: string) => {
  //set the initial state
  const [result, setResult] = useState<Service<CharacterDetails>>({
    status: "loading",
  });
  useEffect(() => {
    const url = `${characterApiEndPoint}/${characterId}`;
    //API function to get the character details
    const fetchCharacterDetailsApi = async () => {
      try {
        // Fetch data from REST API
        const characterResponse = await (await fetch(url)).json();
        //Set the status to error in case of failure
        if (characterResponse.error) {
          setResult({
            status: "error",
            error: {
              name: "character-detail-error",
              message: "Unable to fetch character details",
            },
          });
        } else {
          //Reduce the episode array to single endpioint url with multiple episodes id in slug to get all the episdode details
          const epiSodeUrl: string = characterResponse.episode.reduce(
            (acc: string, url: string) => {
              return `${acc},${url.split(/[\s/]+/).pop()}`;
            }
          );
          //Handle the Episode API call based on the total episodes of the character
          const episodeResponse =
            characterResponse.episode.length !== 0
              ? await (await fetch(epiSodeUrl)).json()
              : [];
          if (episodeResponse.error) {
            setResult({
              status: "error",
              error: {
                name: "episode-detail-error",
                message: "Unable to fetch episode details",
              },
            });
          } else {
            //Handling for Single Episode response as it is returned as object
            const episodeDetails =
              characterResponse.episode.length === 1
                ? [
                    {
                      name: episodeResponse.name,
                      airedDate: episodeResponse.air_date,
                      season: episodeResponse.episode,
                    },
                  ]
                : episodeResponse.map((data: any) => {
                    return {
                      name: data.name,
                      airedDate: data.air_date,
                      season: data.episode,
                    };
                  });
            //Process the response data
            setResult({
              status: "loaded",
              payload: {
                name: characterResponse.name,
                imageUrl: characterResponse.image,
                lastLocation: characterResponse.location.name,
                totalEpisodes: characterResponse.episode.length,
                characterId: characterResponse.id,
                species: characterResponse.species,
                status: characterResponse.status,
                type: characterResponse.type,
                gender: characterResponse.gender,
                origin: characterResponse.origin.name,
                episode: episodeDetails,
                createdAt: characterResponse.created,
              },
            });
          }
        }
      } catch (error) {
        //Update the state to error
        setResult({ status: "error", error });
      }
    };

    // Call async function
    fetchCharacterDetailsApi();
  }, []);
  return result;
};

export default useCharacterDetailService;
