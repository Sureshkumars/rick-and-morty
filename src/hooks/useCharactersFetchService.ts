import { useEffect, useState } from "react";
import { CharacterSearchResults } from "../types/character";
import { Service } from "../types/service";
import { CHARACTER_API_ENDPOINT } from "../constants/constants";

const characterApiEndPoint = CHARACTER_API_ENDPOINT;

const useCharactersFetchService = (
  searchTerm: string,
  page?: number,
  showAll: boolean = false
) => {
  //set the initial state
  const [result, setResult] = useState<Service<CharacterSearchResults>>({
    status: "loading",
  });
  useEffect(() => {
    //When user visits for first time, fetch service won't get called as per our usecase

    if (!searchTerm && !showAll) {
      return;
    }

    //Toggles between searching for all characters or filter based on input search term
    let url =
      searchTerm === ""
        ? `${characterApiEndPoint}/?page=${page}`
        : `${characterApiEndPoint}/?name=${searchTerm}&page=${page}`;
    // Define asynchronous function
    const fetchCharactersApi = async () => {
      try {
        // Fetch data from REST API
        const response = await (await fetch(url)).json();
        //When no characters found API sends response with error key
        if (response.error) {
          setResult({
            status: "loaded",
            payload: {
              totalCount: 0,
              totalPages: 0,
              characters: [],
            },
          });
        } else {
          //Process the response data to according to our defined character types
          setResult({
            status: "loaded",
            payload: {
              totalCount: response.info.count,
              totalPages: response.info.pages,
              characters: response.results.map((data: any) => {
                return {
                  name: data.name,
                  imageUrl: data.image,
                  lastLocation: data.location.name,
                  totalEpisodes: data.episode.length,
                  characterId: data.id,
                  species: data.species,
                };
              }),
            },
          });
        }
      } catch (error) {
        //Catches error and update the return value
        setResult({ status: "error", error });
      }
    };

    // Call async function
    fetchCharactersApi();
  }, [searchTerm, page, showAll]);
  return result;
};

export default useCharactersFetchService;
