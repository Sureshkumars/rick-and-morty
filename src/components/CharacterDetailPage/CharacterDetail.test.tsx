import React from "react";
import { CharacterDetailPage } from "./CharacterDetailPage";
import { render } from "@testing-library/react";
import { MemoryRouter, Route } from "react-router-dom";
import * as hooks from "../../hooks/useCharacterDetailsService";

describe("CharacterDetailPage", () => {
  // Mocking methods which are not implemented in JSDOM
  // https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

  it("Sucessfull Character Details Page Fetch snapshot", () => {
    jest.spyOn(hooks, "default").mockImplementation(() => ({
      status: "loaded",
      payload: {
        name: "Rick Sanchez",
        imageUrl: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
        lastLocation: "Earth (Replacement Dimension)",
        totalEpisodes: 2,
        characterId: 1,
        species: "Human",
        status: "Alive",
        type: "",
        gender: "Male",
        origin: "Earth (C-137)",
        episode: [
          {
            airedDate: "December 9, 2013",
            name: "Sample Episode",
            season: "S01E02",
          },
          {
            airedDate: "December 9, 2013",
            name: "Lawnmower Dog",
            season: "S01E02",
          },
        ],
        createdAt: "2017-11-04T18:48:46.250Z",
      },
    }));

    const { asFragment }: any = render(
      <MemoryRouter initialEntries={["/1"]}>
        <Route exact path="/:characterId">
          <CharacterDetailPage />
        </Route>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
