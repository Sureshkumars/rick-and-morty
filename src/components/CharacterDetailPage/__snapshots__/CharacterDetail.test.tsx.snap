// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailPage Sucessfull Character Details Page Fetch snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    <div
      class="chakra-stack css-1cvi3v3"
    >
      <img
        alt="Rick Sanchez"
        class="chakra-image__placeholder css-cl5oj8"
        src="/images/character-fallback.jpeg"
      />
      <div
        class="css-0"
      >
        <h2
          class="chakra-heading css-1itv5e3"
        >
          Rick Sanchez
        </h2>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                species
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                Human
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                Gender
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                Male
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                Status
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                Alive
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                Origin Location
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                Earth (C-137)
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                Last Known Location
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                Earth (Replacement Dimension)
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <div
            class="chakra-stat css-0"
          >
            <dl>
              <dt
                class="chakra-stat__label css-0"
              >
                Total Episodes
              </dt>
              <dd
                class="chakra-stat__number css-fvxuh5"
              >
                2
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="css-ll0rx5"
        >
          <ul
            class="css-c5ckgy"
            role="list"
          >
            <div
              class="chakra-stat css-0"
            >
              <dl>
                <dt
                  class="chakra-stat__label css-0"
                >
                  Episode Details
                </dt>
              </dl>
            </div>
            <li
              class="css-0"
            >
              <svg
                class="chakra-icon css-1v3wlva"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="presentation"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                />
              </svg>
              Sample Episode - S01E02 - was aired at December 9, 2013 
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1c5urqo"
                role="separator"
              />
            </li>
            <li
              class="css-0"
            >
              <svg
                class="chakra-icon css-1v3wlva"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="presentation"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                />
              </svg>
              Lawnmower Dog - S01E02 - was aired at December 9, 2013 
              <hr
                aria-orientation="horizontal"
                class="chakra-divider css-1c5urqo"
                role="separator"
              />
            </li>
          </ul>
        </div>
        <button
          aria-label="buy-merchandise"
          class="chakra-button css-jaqb0y"
          type="button"
        >
          Buy Merchandise
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
